File: drake_ros_bazel_installed/apps/inquirer.cc
#include <functional>
#include <memory>

#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/client.hpp>
#include <rclcpp_action/create_client.hpp>

#include "apps_msgs/msg/status.hpp"
#include "common_msgs/srv/query.hpp"
#include "common_msgs/action/do.hpp"

using namespace std::chrono_literals;

namespace apps {

class Inquirer : public rclcpp::Node {
public:
  Inquirer() : Node("inquirer") {
    using namespace std::placeholders;
    status_sub_ =
      this->create_subscription<apps_msgs::msg::Status>(
        "status", rclcpp::QoS(rclcpp::KeepLast(1)),
        std::bind(&Inquirer::on_status, this, _1));

    oracle_query_client_ = this->create_client<common_msgs::srv::Query>("query");

    oracle_action_client_ = rclcpp_action::create_client<common_msgs::action::Do>(this, "do");

    inquire_timer_ = this->create_wall_timer(5s, std::bind(&Inquirer::inquire, this));
  }

private:
  using QueryClient = rclcpp::Client<common_msgs::srv::Query>;

  void handle_query_response(QueryClient::SharedFuture future) const {
    RCLCPP_INFO(this->get_logger(), "oracle said: %s", future.get()->reply.c_str());
  }

  using ActionGoalHandle = rclcpp_action::ClientGoalHandle<common_msgs::action::Do>;
  void handle_rite_request_response(const std::shared_ptr<ActionGoalHandle> handle) const {
    if (!handle) {
      RCLCPP_ERROR(this->get_logger(), "oracle rejected rite request");
    } else {
      RCLCPP_INFO(this->get_logger(), "oracle rite in progress");
    }
  }

  void handle_rite_feedback(
    std::shared_ptr<ActionGoalHandle> handle,
    const std::shared_ptr<const ActionGoalHandle::Feedback> feedback) const
  {
    RCLCPP_INFO(this->get_logger(), "oracle is %s", feedback->message.c_str());
  }

  void handle_rite_result(const ActionGoalHandle::WrappedResult& result) const
  {
    switch (result.code) {
      case rclcpp_action::ResultCode::SUCCEEDED:
        RCLCPP_INFO(this->get_logger(), "oracle rite is complete");
        break;
      case rclcpp_action::ResultCode::ABORTED:
        RCLCPP_ERROR(this->get_logger(),
                     "oracle rite aborted due to %s",
                     result.result->reason.c_str());
        break;
      case rclcpp_action::ResultCode::CANCELED:
        RCLCPP_WARN(this->get_logger(), "oracle rite was cancelled");
        break;
      default:
        RCLCPP_ERROR(this->get_logger(), "oracle rite outcome unknown");
        break;
    }
  }

  void inquire() const {
    using namespace std::placeholders;

    auto request = std::make_shared<common_msgs::srv::Query::Request>();
    request->query = "how's it going?";
    if (oracle_query_client_->service_is_ready()) {
      RCLCPP_INFO(this->get_logger(), "oracle, %s", request->query.c_str());
      oracle_query_client_->async_send_request(
        request, std::bind(&Inquirer::handle_query_response, this, _1));
    } else {
      RCLCPP_WARN(this->get_logger(), "oracle not available for queries");
    }

    if (oracle_action_client_->action_server_is_ready()) {
      rclcpp_action::Client<common_msgs::action::Do>::SendGoalOptions options;
      options.goal_response_callback =
        std::bind(&Inquirer::handle_rite_request_response, this, _1);
      options.feedback_callback =
        std::bind(&Inquirer::handle_rite_feedback, this, _1, _2);
      options.result_callback =
        std::bind(&Inquirer::handle_rite_result, this, _1);
      common_msgs::action::Do::Goal goal;
      goal.action = "rite";
      goal.period = rclcpp::Duration::from_seconds(0.1);
      goal.timeout = rclcpp::Duration::from_seconds(1.0);
      oracle_action_client_->async_send_goal(goal, options);
    } else {
      RCLCPP_WARN(this->get_logger(), "oracle not available for actions");
    }
  }

  void on_status(const apps_msgs::msg::Status & msg) const {
    RCLCPP_INFO(
      get_logger(), "%s status (%lu): %s",
      msg.origin.c_str(), msg.status.sequence_id,
      msg.status.message.c_str());
  }

  std::shared_ptr<rclcpp::Subscription<apps_msgs::msg::Status>> status_sub_;
  std::shared_ptr<rclcpp::Client<common_msgs::srv::Query>> oracle_query_client_;
  std::shared_ptr<rclcpp_action::Client<common_msgs::action::Do>> oracle_action_client_;
  std::shared_ptr<rclcpp::TimerBase> inquire_timer_;
};

}  // namespace apps

int main(int argc, char ** argv) {
  rclcpp::init(argc, argv);

  rclcpp::spin(std::make_shared<apps::Inquirer>());

  rclcpp::shutdown();
}

File: drake_ros_bazel_installed/apps/inquirer.py
#!/usr/bin/env python3

import action_msgs.msg
import rclpy
import rclpy.action
import rclpy.duration
import rclpy.node

import apps_msgs.msg
import common_msgs.action
import common_msgs.srv


class Inquirer(rclpy.node.Node):

    def __init__(self):
        super().__init__('inquirer')
        self._status_sub = self.create_subscription(
            apps_msgs.msg.Status, 'status', self._on_status, 1)
        self._query_client = self.create_client(common_msgs.srv.Query, 'query')
        self._action_client = rclpy.action.ActionClient(
            self, common_msgs.action.Do, 'do')
        self._inquire_timer = self.create_timer(5.0, self.inquire)

    def _on_status(self, msg):
        self.get_logger().info('{} status ({}): {}'.format(
            msg.origin, msg.status.sequence_id, msg.status.message
        ))

    def _handle_reply(self, future):
        self.get_logger().info('oracle said: ' + future.result().reply)

    def _handle_rite_request_response(self, future):
        handle = future.result()
        if not handle.accepted:
            self.get_logger().error('oracle rejected rite request')
            return
        self.get_logger().info('oracle rite in progress')
        future = handle.get_result_async()
        future.add_done_callback(self._handle_rite_result)

    def _handle_rite_feedback(self, msg):
        self.get_logger().info('oracle is ' + msg.feedback.message)

    def _handle_rite_result(self, future):
        result = future.result()
        if result.status == action_msgs.msg.GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info('oracle rite is complete')
        elif result.status == action_msgs.msg.GoalStatus.STATUS_ABORTED:
            self.get_logger().error(
                'oracle rite aborted due to ' + result.result.reason
            )
        elif result.status == action_msgs.msg.GoalStatus.STATUS_CANCELED:
            self.get_logger().error('oracle rite was cancelled')
        else:
            self.get_logger().error('oracle rite status unknown')

    def inquire(self):
        if self._query_client.service_is_ready():
            request = common_msgs.srv.Query.Request()
            request.query = "how's it going?"
            self.get_logger().info('oracle, ' + request.query)
            future = self._query_client.call_async(request)
            future.add_done_callback(self._handle_reply)
        else:
            self.get_logger().warning('oracle not available for queries')
        if self._action_client.server_is_ready():
            goal = common_msgs.action.Do.Goal()
            goal.action = 'rite'
            goal.period = rclpy.duration.Duration(seconds=0.1).to_msg()
            goal.timeout = rclpy.duration.Duration(seconds=1.0).to_msg()
            future = self._action_client.send_goal_async(
                goal, feedback_callback=self._handle_rite_feedback)
            future.add_done_callback(self._handle_rite_request_response)
        else:
            self.get_logger().warning('oracle not available for actions')

def main():
    rclpy.init()

    try:
        rclpy.spin(Inquirer())
    except KeyboardInterrupt:
        pass
    finally:
        rclpy.shutdown()


if __name__ == '__main__':
    main()
